### PR TITLE
Imagehub refatcor.

### PR DESCRIPTION
**Pull request by lingdie**

### CODE REVIEW

File: controllers/imagehub/controllers/image_controller.go [hunk 1]
```diff
@@ -61,66 +60,52 @@ type ImageReconciler struct {
...
-func (r *ImageReconciler) Delete(ctx context.Context, req ctrl.Request, gvk schema.GroupVersionKind, obj client.Object) error {
-	r.Logger.V(1).Info("delete reconcile controller image", "request", req)
-	// get img from obj
-	img := &imagehubv1.Image{}
-	err := convert.JsonConvert(obj, img)
-	if err != nil {
-		r.Logger.V(2).Info("error in image json convert", "namespace/name", obj.GetNamespace()+"/"+obj.GetName())
-		return err
+func (r *ImageReconciler) doFinalizer(ctx context.Context, obj client.Object) error {
+	r.Logger.V(1).Info("delete reconcile controller image", "request", client.ObjectKeyFromObject(obj))
```
Comment (zzjin) [inline controllers/imagehub/controllers/image_controller.go:97]: this log will internally using json convert, this seems too heavy for just logs.

File: controllers/imagehub/controllers/image_controller.go [hunk 2]
```diff
@@ -61,66 +60,52 @@ type ImageReconciler struct {
...
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return r.doReconcile(ctx, img)
 	}
+	return ctrl.Result{}, errors.New("reconcile error from Finalizer")
+}
 
+func (r *ImageReconciler) doReconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	r.Logger.V(1).Info("update reconcile controller image", "request", client.ObjectKeyFromObject(obj))
```
Comment (zzjin) [inline controllers/imagehub/controllers/image_controller.go:82]: This one too.

File: controllers/imagehub/controllers/image_controller.go (hunk 1 above)
Comment (lingdie) [reply controllers/imagehub/controllers/image_controller.go:97]: how to solve it?

Comment (lingdie) [reply controllers/imagehub/controllers/image_controller.go:97]: https://github.com/labring/sealos/blob/70b4aaa70ca53dcb158f9a357c32aeaa05b4fb98/controllers/user/controllers/user_controller.go#L134

File: controllers/imagehub/api/v1/organization_types.go [hunk 3]
```diff
@@ -50,12 +48,16 @@ func (n OrgName) ToMetaName() string {
...
-type OrgInfo OrganizationSpec
+type OrgInfo OrganizationStatus
 
 // OrganizationStatus defines the observed state of Organization
 type OrganizationStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+	//+kubebuilder:validation:Required
+	//+kubebuilder:validation:MaxLength=1024
+	Name  string     `json:"name,omitempty"` //e.g: "libring"
```
Comment (fanux) [inline controllers/imagehub/api/v1/organization_types.go:59]: libring->labring

File: controllers/imagehub/controllers/data_helper.go [hunk 4]
```diff
@@ -14,6 +14,8 @@ type DataHelper struct {
 	logr.Logger
 }
 
+var ErrNoMatch = errors.New("NoMatch")
```
Comment (fanux) [inline controllers/imagehub/controllers/data_helper.go:17]: NotMatch?

File: controllers/imagehub/controllers/image_controller.go [hunk 5]
```diff
@@ -61,159 +58,76 @@ type ImageReconciler struct {
...
-	for _, t := range repo.Spec.Tags {
-		if t.Name != img.Spec.Name.GetTag() {
-			spec.Tags = append(spec.Tags, t)
+	// create repo and SetControllerReference.
+	repo := &imagehubv1.Repository{}
+	repo.Name = img.Spec.Name.ToRepoName().ToMetaName()
+	err := r.Get(ctx, client.ObjectKeyFromObject(repo), repo)
+	if err != nil && apierrors.IsNotFound(err) {
+		repo.Spec.Name = img.Spec.Name.ToRepoName()
+		if err := r.Create(ctx, repo); err != nil {
```
Comment (zzjin) [inline controllers/imagehub/controllers/image_controller.go:91]: If create image need to create repo, but create action failed, code continues to run?

Comment (lingdie) [reply controllers/imagehub/controllers/image_controller.go:91]: returned.

**Review by zzjin — APPROVED** [review]

File: controllers/imagehub/controllers/data_helper.go [hunk 6]
```diff
@@ -108,19 +110,25 @@ func (r *DataHelper) genFulldataByImageName(ctx context.Context, n imagehubv1.Im
 	fd := imagehubv1.FullData{}
 
 	orgInfo, err := r.getOrgInfoByOrgName(ctx, n.ToOrgName())
-	if err != nil {
+	if err == ErrNotMatch {
+		r.Logger.V(2).Info("getOrgInfoByOrgName", "err:", err.Error())
```
Comment (fanux) [inline controllers/imagehub/controllers/data_helper.go:114]: failed to get origination info by name: %s, name

File: controllers/imagehub/controllers/image_controller.go [hunk 7]
```diff
@@ -61,159 +58,77 @@ type ImageReconciler struct {
...
-	}
-
-	// todo try to delete image in hub.sealos.io registry
-
-	// get repo by img spec name
-	repo, err := r.db.getRepoByRepoName(ctx, img.Spec.Name.ToRepoName())
-	if err != nil {
-		r.Logger.V(2).Info("error in image getRepoByRepoName, not found repo by lable", "err:", err.Error())
-		return err
+func (r *ImageReconciler) doReconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
```
Comment (fanux) [inline controllers/imagehub/controllers/image_controller.go:79]: doReconcile -> reconcile

Comment (lingdie) [reply controllers/imagehub/controllers/image_controller.go:79]: done at [pr 2233](https://github.com/labring/sealos/pull/2233)

File: controllers/imagehub/controllers/data_helper.go (hunk 6 above)
Comment (lingdie) [reply controllers/imagehub/controllers/data_helper.go:114]: done at [pr 2233](https://github.com/labring/sealos/pull/2233)

File: controllers/imagehub/api/v1/organization_types.go (hunk 3 above)
Comment (lingdie) [reply controllers/imagehub/api/v1/organization_types.go]: done.

**Pull request merged**